controller/runner.go

```diff
@@ -0,0 +1,186 @@
+type ServerRunner struct {
+	log  *logp.Logger
+	cfg  config.ShipperConfig
+	once sync.Once
+	mu   sync.Mutex
+
+	server     *grpc.Server
+	shipper    server.ShipperServer
+	queue      *queue.Queue
+	monitoring *monitoring.QueueMonitor
+	out        Output
+}
+
+// NewServerRunner creates a new runner that starts and stops the server.
+// Consumers are expected to run `Close` to release all resources created
+// by a successful result from this function even without starting the server.
+func NewServerRunner(cfg config.ShipperConfig) (r *ServerRunner, err error) {
+	r = &ServerRunner{
+		log: logp.L(),
+		cfg: cfg,
+	}
+	// in case of an initialization error we must clean up all created resources
+	defer func() {
+		if err != nil {
+			// this will account for partial initialization in case of an error, so there are no leaks
+			r.Close()
+		}
+	}()
+
+	r.log.Debug("initializing the queue...")
+	r.queue, err = queue.New(cfg.Queue)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't create queue: %w", err)
+	}
+	r.log.Debug("queue was initialized.")
+
+	r.log.Debug("initializing monitoring ...")
+	r.monitoring, err = monitoring.NewFromConfig(cfg.Monitor, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing output monitor: %w", err)
+	}
+	r.monitoring.Watch()
+	r.log.Debug("monitoring is ready.")
+
+	r.log.Debug("initializing the output...")
+	// TODO replace with the real output based on the config, Console is hard-coded for now
+	r.out = output.NewConsole(r.queue)
+	r.out.Start()
+	r.log.Debug("output was initialized.")
+
+	r.log.Debug("initializing the gRPC server...")
+	var opts []grpc.ServerOption
+	if cfg.Server.TLS {
+		creds, err := credentials.NewServerTLSFromFile(cfg.Server.Cert, cfg.Server.Key)
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate credentials %w", err)
+		}
+		opts = []grpc.ServerOption{grpc.Creds(creds)}
+	}
+	r.server = grpc.NewServer(opts...)
+	r.shipper, err = server.NewShipperServer(cfg.Server, r.queue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialise the server: %w", err)
+	}
+	pb.RegisterProducerServer(r.server, r.shipper)
+	r.log.Debug("gRPC server initialized.")
+
+	return r, nil
+}
+
+// Start runs the shipper server according to the set configuration.
+// The server stops after calling `Close`, this function blocks until then.
+func (r *ServerRunner) Start() (err error) {
+	r.mu.Lock()
+	if r.server == nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to start a server runner that was previously closed")
+	}
+
+	addr := fmt.Sprintf("localhost:%d", r.cfg.Server.Port)
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		r.mu.Unlock()
+		return fmt.Errorf("failed to listen on %s: %w", addr, err)
+	}
+
+	// Testing that the server is running and only then unlock the mutex.
+	// Otherwise if `Close` is called at the same time as `Start` it causes race condition.
+	go func() {
+		defer r.mu.Unlock()
```

Comment:
Is that the only thing this mutex is used for? If so, you might want to rename it, I was a tad confused by the mutex itself.

Alternatively, depending on the where the race is in `Close()` you could also use a WaitGroup, or just an atomic flag that can tell the `Close()` what shutdown components to avoid

Reply:
The reason this mutex exists is that the controller tests call `Close` before `Start` finishes. Which is theoretically possible with the control protocol, I guess. I'll rename it to `startMutex`.